File: extract_fortification_duas.py
#!/usr/bin/env python3
"""
Fortification of the Muslim (Hisnul Muslim) - Dua Extractor
============================================================

Extracts duas from the PDF and formats them for Prisma database import.

Output: fortification_duas.json
"""

import PyPDF2
import json
import re
from pathlib import Path
from typing import List, Dict, Optional

# PDF Path
PDF_PATH = "quran-hadith-app/quran-hadith-app/data/quran-nlp/data/Fortification of the Muslim (from the Quran and Sunnah) (Sa'id ibn Ali ibn Wahf al-Qahtani) (Z-Library).pdf"

# Category mappings (Hisnul Muslim categories)
CATEGORY_PATTERNS = {
    "Morning": ["morning", "awaking", "wake up", "after waking"],
    "Evening": ["evening", "before sleeping", "sleep", "going to bed"],
    "Prayer": ["prayer", "salah", "prostration", "sujud", "ruku", "after prayer"],
    "Quran": ["quran", "recit"],
    "Mosque": ["mosque", "masjid", "entering", "leaving"],
    "Adhan": ["adhan", "azan", "call to prayer"],
    "Travel": ["travel", "journey", "vehicle", "riding"],
    "Food": ["food", "eating", "drink", "meal"],
    "Distress": ["distress", "anxiety", "worry", "grief", "sorrow", "difficulty"],
    "Forgiveness": ["forgiveness", "repentance", "sin", "istighfar"],
    "Protection": ["protection", "evil", "satan", "shaytan", "refuge"],
    "Family": ["family", "children", "spouse", "marriage"],
    "Health": ["sick", "illness", "health", "disease", "pain"],
    "Rain": ["rain", "thunder", "lightning"],
    "General": ["general", "various", "miscellaneous"]
}

def extract_text_from_pdf(pdf_path: str) -> str:
    """Extract all text from PDF."""
    print(f"📄 Reading PDF: {pdf_path}")

    with open(pdf_path, 'rb') as file:
        reader = PyPDF2.PdfReader(file)
        text = ""

        total_pages = len(reader.pages)
        print(f"   Total pages: {total_pages}")

        for i, page in enumerate(reader.pages):
            text += page.extract_text()
            if (i + 1) % 10 == 0:
                print(f"   Processed {i + 1}/{total_pages} pages...")

        print(f"   ✅ Extracted {len(text)} characters\n")
        return text

def detect_category(text: str) -> str:
    """Detect dua category from context."""
    text_lower = text.lower()

    for category, patterns in CATEGORY_PATTERNS.items():
        for pattern in patterns:
            if pattern in text_lower:
                return category

    return "General"

def is_arabic(text: str) -> bool:
    """Check if text contains Arabic characters."""
    arabic_pattern = re.compile(r'[\u0600-\u06FF]+')
    return bool(arabic_pattern.search(text))

def clean_text(text: str) -> str:
    """Clean extracted text."""
    # Remove excessive whitespace
    text = re.sub(r'\s+', ' ', text)
    # Remove page numbers
    text = re.sub(r'\b\d+\s*$', '', text)
    return text.strip()

def extract_reference(text: str) -> Optional[str]:
    """Extract hadith reference from text."""
    # Common patterns for references
    patterns = [
        r'(?:Bukhari|Muslim|Abu Dawud|Tirmidhi|Nasa\'i|Ibn Majah)[\s:]+\d+',
        r'Sahih\s+(?:al-)?Bukhari\s+\d+',
        r'Sahih\s+Muslim\s+\d+',
        r'Sunan\s+\w+\s+\d+',
        r'Quran\s+\d+:\d+',
        r'Al-Quran\s+\d+:\d+',
    ]

    for pattern in patterns:
        match = re.search(pattern, text, re.IGNORECASE)
        if match:
            return match.group(0)

    return None

def split_into_duas(text: str) -> List[Dict]:
    """Parse text into individual duas."""
    print("🔍 Parsing duas from text...")

    duas = []

    # Split by common section markers
    sections = re.split(r'\n\s*\n|\d+\.\s+', text)

    current_category = "General"
    dua_number = 1

    for section in sections:
        section = clean_text(section)

        if len(section) < 20:  # Skip very short sections
            continue

        # Check if this is a category header
        if len(section) < 100 and any(cat.lower() in section.lower() for cat in CATEGORY_PATTERNS.keys()):
            current_category = detect_category(section)
            continue

        # Split section into lines
        lines = [line.strip() for line in section.split('\n') if line.strip()]

        if len(lines) < 2:
            continue

        # Try to identify Arabic text, translation, and transliteration
        arabic_text = None
        english_text = None
        transliteration = None
        reference = extract_reference(section)

        for line in lines:
            if is_arabic(line) and not arabic_text:
                arabic_text = line
            elif line.startswith('(') or '[' in line:
                # Likely transliteration (often in parentheses)
                if not transliteration:
                    transliteration = line.strip('()[]')
            elif len(line) > 20 and not english_text:
                english_text = line

        # Create dua entry if we have at least Arabic or English
        if arabic_text or english_text:
            dua = {
                "number": dua_number,
                "category": current_category,
                "title": f"{current_category} Dua #{dua_number}",
                "arabic": arabic_text or "",
                "english": english_text or "",
                "transliteration": transliteration or "",
                "reference": reference or "",
                "occasion": current_category
            }

            duas.append(dua)
            dua_number += 1

            if dua_number % 10 == 0:
                print(f"   Extracted {dua_number} duas...")

    print(f"   ✅ Total duas extracted: {len(duas)}\n")
    return duas

def generate_prisma_format(duas: List[Dict]) -> Dict:
    """Format duas for Prisma import."""
    print("📦 Formatting for Prisma...")

    # Group by category
    categories = {}
    for dua in duas:
        cat = dua['category']
        if cat not in categories:
            categories[cat] = []
        categories[cat].append(dua)

    print(f"   Categories found: {len(categories)}")
    for cat, items in categories.items():
        print(f"   - {cat}: {len(items)} duas")

    # Create output structure
    output = {
        "source": "Fortification of the Muslim (Hisnul Muslim)",
        "author": "Sa'id ibn Ali ibn Wahf al-Qahtani",
        "total_duas": len(duas),
        "categories": list(categories.keys()),
        "duas": []
    }

    # Format each dua
    for dua in duas:
        formatted = {
            "title": dua['title'],
            "titleArabic": None,
            "textArabic": dua['arabic'],
            "textEnglish": dua['english'],
            "transliteration": dua['transliteration'] if dua['transliteration'] else None,
            "reference": dua['reference'] if dua['reference'] else None,
            "category": dua['category'],
            "tags": f"{dua['category']},Daily Duas,Hisnul Muslim",
            "benefits": None,
            "occasion": dua['occasion']
        }
        output["duas"].append(formatted)

    return output

def save_json(data: Dict, output_file: str):
    """Save data to JSON file."""
    print(f"💾 Saving to {output_file}...")

    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(data, f, ensure_ascii=False, indent=2)

    print(f"   ✅ Saved successfully!\n")

def main():
    """Main execution."""
    print("\n" + "="*70)
    print("FORTIFICATION OF THE MUSLIM - DUA EXTRACTOR")
    print("="*70 + "\n")

    # Check if PDF exists
    pdf_path = Path(PDF_PATH)
    if not pdf_path.exists():
        print(f"❌ Error: PDF not found at {PDF_PATH}")
        print(f"   Please ensure the file exists at this location.")
        return

    try:
        # Extract text from PDF
        text = extract_text_from_pdf(PDF_PATH)

        # Parse duas
        duas = split_into_duas(text)

        if not duas:
            print("⚠️  Warning: No duas extracted. PDF might be image-based or encrypted.")
            print("   Try using an OCR tool first.")
            return

        # Format for Prisma
        output = generate_prisma_format(duas)

        # Save to JSON
        output_file = "fortification_duas.json"
        save_json(output, output_file)

        # Summary
        print("="*70)
        print("EXTRACTION COMPLETE!")
        print("="*70)
        print(f"\n📊 Summary:")
        print(f"   Total Duas: {len(duas)}")
        print(f"   Categories: {len(output['categories'])}")
        print(f"   Output File: {output_file}")
        print(f"\n📝 Next Steps:")
        print(f"   1. Review the JSON file: {output_file}")
        print(f"   2. Import into database using Prisma")
        print(f"   3. Verify duas are correctly formatted")
        print("\n" + "="*70 + "\n")

    except Exception as e:
        print(f"\n❌ Error: {e}")
        import traceback
        traceback.print_exc()

if __name__ == "__main__":
    main()

File: quran-hadith-app/quran-hadith-app/scripts/extract-fortification-duas.ts
/**
 * Fortification of the Muslim (Hisnul Muslim) - Dua Extractor
 * =============================================================
 *
 * Extracts duas from the PDF and formats them for Prisma database import.
 */

import * as fs from 'fs';
import * as path from 'path';
import { glob } from 'glob';
// @ts-ignore
import PDFParse from 'pdf-parse';

interface Dua {
  number: number;
  category: string;
  title: string;
  titleArabic: string | null;
  textArabic: string;
  textEnglish: string;
  transliteration: string | null;
  reference: string | null;
  tags: string;
  occasion: string;
  benefits: string | null;
}

// Category patterns for detection
const CATEGORY_PATTERNS: Record<string, string[]> = {
  Morning: ['morning', 'awaking', 'wake up', 'after waking'],
  Evening: ['evening', 'before sleeping', 'sleep', 'going to bed'],
  Prayer: ['prayer', 'salah', 'prostration', 'sujud', 'ruku', 'after prayer'],
  Quran: ['quran', 'recit'],
  Mosque: ['mosque', 'masjid', 'entering', 'leaving'],
  Adhan: ['adhan', 'azan', 'call to prayer'],
  Travel: ['travel', 'journey', 'vehicle', 'riding'],
  Food: ['food', 'eating', 'drink', 'meal'],
  Distress: ['distress', 'anxiety', 'worry', 'grief', 'sorrow', 'difficulty'],
  Forgiveness: ['forgiveness', 'repentance', 'sin', 'istighfar'],
  Protection: ['protection', 'evil', 'satan', 'shaytan', 'refuge'],
  Family: ['family', 'children', 'spouse', 'marriage'],
  Health: ['sick', 'illness', 'health', 'disease', 'pain'],
  Rain: ['rain', 'thunder', 'lightning'],
  General: ['general', 'various', 'miscellaneous']
};

function isArabic(text: string): boolean {
  const arabicPattern = /[\u0600-\u06FF]/;
  return arabicPattern.test(text);
}

function detectCategory(text: string): string {
  const textLower = text.toLowerCase();

  for (const [category, patterns] of Object.entries(CATEGORY_PATTERNS)) {
    for (const pattern of patterns) {
      if (textLower.includes(pattern)) {
        return category;
      }
    }
  }

  return 'General';
}

function extractReference(text: string): string | null {
  const patterns = [
    /(?:Bukhari|Muslim|Abu Dawud|Tirmidhi|Nasa'i|Ibn Majah)[\s:]+\d+/i,
    /Sahih\s+(?:al-)?Bukhari\s+\d+/i,
    /Sahih\s+Muslim\s+\d+/i,
    /Sunan\s+\w+\s+\d+/i,
    /Quran\s+\d+:\d+/i,
    /Al-Quran\s+\d+:\d+/i,
  ];

  for (const pattern of patterns) {
    const match = text.match(pattern);
    if (match) {
      return match[0];
    }
  }

  return null;
}

function cleanText(text: string): string {
  return text.replace(/\s+/g, ' ').replace(/\b\d+\s*$/, '').trim();
}

async function extractDuas(pdfPath: string): Promise<Dua[]> {
  console.log('📄 Reading PDF:', pdfPath);

  const dataBuffer = fs.readFileSync(pdfPath);
  const data = await PDFParse(dataBuffer);

  console.log(`   Total pages: ${data.numpages}`);
  console.log(`   Extracted ${data.text.length} characters\n`);

  console.log('🔍 Parsing duas from text...');

  const duas: Dua[] = [];
  const sections = data.text.split(/\n\s*\n|\d+\.\s+/);

  let currentCategory = 'General';
  let duaNumber = 1;

  for (const section of sections) {
    const cleanSection = cleanText(section);

    if (cleanSection.length < 20) continue;

    // Check if category header
    if (cleanSection.length < 100) {
      const detectedCategory = detectCategory(cleanSection);
      if (detectedCategory !== 'General') {
        currentCategory = detectedCategory;
        continue;
      }
    }

    const lines = cleanSection.split('\n').map(l => l.trim()).filter(l => l);

    if (lines.length < 2) continue;

    let arabicText = '';
    let englishText = '';
    let transliteration: string | null = null;
    const reference = extractReference(cleanSection);

    for (const line of lines) {
      if (isArabic(line) && !arabicText) {
        arabicText = line;
      } else if (line.startsWith('(') || line.includes('[')) {
        if (!transliteration) {
          transliteration = line.replace(/[()[\]]/g, '').trim();
        }
      } else if (line.length > 20 && !englishText) {
        englishText = line;
      }
    }

    if (arabicText || englishText) {
      duas.push({
        number: duaNumber,
        category: currentCategory,
        title: `${currentCategory} Dua #${duaNumber}`,
        titleArabic: null,
        textArabic: arabicText,
        textEnglish: englishText,
        transliteration,
        reference,
        tags: `${currentCategory},Daily Duas,Hisnul Muslim`,
        occasion: currentCategory,
        benefits: null
      });

      duaNumber++;

      if (duaNumber % 10 === 0) {
        console.log(`   Extracted ${duaNumber} duas...`);
      }
    }
  }

  console.log(`   ✅ Total duas extracted: ${duas.length}\n`);
  return duas;
}

async function main() {
  console.log('\n' + '='.repeat(70));
  console.log('FORTIFICATION OF THE MUSLIM - DUA EXTRACTOR');
  console.log('='.repeat(70) + '\n');

  // Find the PDF using glob pattern to handle special characters
  console.log('🔍 Searching for PDF file...');
  const searchPattern = path.join(process.cwd(), 'data/duas-pdfs/Fortification*.pdf');
  const files = await glob(searchPattern);

  if (files.length === 0) {
    console.log('❌ Error: PDF not found');
    console.log('   Searched in:', searchPattern);
    return;
  }

  const pdfPath = files[0];
  console.log('   ✅ Found:', path.basename(pdfPath), '\n');

  try {
    const duas = await extractDuas(pdfPath);

    if (duas.length === 0) {
      console.log('⚠️  Warning: No duas extracted. PDF might be image-based or encrypted.');
      console.log('   Try using an OCR tool first.');
      return;
    }

    // Group by category
    const categories = duas.reduce((acc, dua) => {
      if (!acc[dua.category]) {
        acc[dua.category] = [];
      }
      acc[dua.category].push(dua);
      return acc;
    }, {} as Record<string, Dua[]>);

    console.log('📦 Formatting for Prisma...');
    console.log(`   Categories found: ${Object.keys(categories).length}`);
    for (const [cat, items] of Object.entries(categories)) {
      console.log(`   - ${cat}: ${items.length} duas`);
    }

    const output = {
      source: 'Fortification of the Muslim (Hisnul Muslim)',
      author: 'Sa\'id ibn Ali ibn Wahf al-Qahtani',
      total_duas: duas.length,
      categories: Object.keys(categories),
      duas: duas
    };

    // Save to JSON
    const outputFile = path.join(process.cwd(), '../fortification_duas.json');
    fs.writeFileSync(outputFile, JSON.stringify(output, null, 2), 'utf-8');

    console.log(`💾 Saved to ${outputFile}\n`);

    console.log('='.repeat(70));
    console.log('EXTRACTION COMPLETE!');
    console.log('='.repeat(70));
    console.log(`\n📊 Summary:`);
    console.log(`   Total Duas: ${duas.length}`);
    console.log(`   Categories: ${Object.keys(categories).length}`);
    console.log(`   Output File: ${outputFile}`);
    console.log(`\n📝 Next Steps:`);
    console.log(`   1. Review the JSON file`);
    console.log(`   2. Import into database using Prisma`);
    console.log(`   3. Verify duas are correctly formatted`);
    console.log('\n' + '='.repeat(70) + '\n');

  } catch (error) {
    console.error('\n❌ Error:', error);
  }
}

main();
